Separate ws from logic

bitterfly opened the pull request.

Commit:
[backend] Check if only one word.

Don't normalize weights when there is only one word to recommend.

Commit:
Merge branch 'master' of github.com:bitterfly/hatgame

Commit:
[backend] defer wsMutex unlock to avoid deadlock

Commit:
[backend] Seed the random when creating game

Commit:
[backend] Replace Game struct on server

Add a new structure which contains the players' websockets and the game
state so that the websocket part could be separated from the business
logic more easily.

Commit:
[backend] Move WS logic to the server

Remove the wss from the Player struct and add them to the server.
The game now has notification channel and when the state of the game
changes an event is sent to the event channel. The server then handles
those events and sends messages via the players' websockets.

bitterfly committed:
[frontend] Make Id uppercase when json marshalling.

bitterfly committed:
[backend] Close ws when event channel is closed

Instead of closing the websockets when GameEnd channel signals
close them when event channels is closed. There can still be
unsent (tick) events when the game has ended so the websockets must be
closed after sending all queued events.

bitterfly committed:
[backend] Remove ws from game

Delete the unnecessary creating of a map with websockets and the import
statement.

bitterfly committed:
[backend] Ignore some game field when json marshalling

The clients only need to know the game info and the users in the game.
All else is ignored.

bitterfly committed:
[backend] Initialize events channel

This was a bug.

bitterfly committed:
[backend] Change return types in some functions

AddWords no longer returns the message bytes but returns whether the
operation was successful (and an error if present).
MakeTurn no longer returns an error.

bitterfly committed:
[backend] Notify the storyteller after the last tick.

Instead of notifying the storyteller in MakeTurn with timerDone channel notify the
storyteller in the tick function since a race condition could occure
between MakeTurn and tick and the storyteller event can be sent before
the tick event.

bitterfly committed:
[backend] Call only game methods from HandleMessage.

Move the notify functions in the game functions.
Also make errors part of the game.

bitterfly committed:
[backend] move HandleMessage to server

Also remove process and make its functions public Game functions.

bitterfly committed:
[backend] Move message and game into separate packages

bitterfly committed:
[backend] Some stylistic changes

stoyaneft commented on hatgame-backend/server/game/game.go:
```diff
@@ -0,0 +1,348 @@
+package game
```
My idea was that the `game` package will be outside `server` folder as its sibling, not as a child.

stoyaneft commented on hatgame-backend/server/message/message.go:
```diff
@@ -0,0 +1,21 @@
+package message
```
There could be `message.go` file next to the `server.go` one. No need of separate `message` folder.

bitterfly replied:
Since game has to send messages on events, game.go imports message. But server.go imports game. If i move message on the same level as server.go it will be in the server package and that will result in an import cycle (which I'm not sure how to resolve). Also if I had to create an Error message in game this will be written as server.Error and I just thought that message,Error reads better that's why I separated it into a message package.

bitterfly replied:
Done

stoyaneft commented on hatgame-backend/server/server.go:
```diff
@@ -362,59 +390,67 @@ func (s *Server) handleHost(w http.ResponseWriter, r *http.Request) {
-	}
-	s.listen(ws, game, payload.Id)
+	go func(g *Game) {
+		for event := range g.State.Events {
```
You can use directly currentGame.State.Events from the closure, no need to pass it as explicit argument.

stoyaneft commented on hatgame-backend/server/server.go:
```diff
@@ -362,59 +390,67 @@ func (s *Server) handleHost(w http.ResponseWriter, r *http.Request) {
+		}
+	}(s.Games[gameID])
+
+	game.NotifyGameInfo(currentGame)
```
You are using `game.NotifyGameInfo` only for putting a new event in the game.Events channel. This doesn't change the game state at all. In this case it's better to directly send the ws message. It's a bit quirk otherwise.

stoyaneft commented on hatgame-backend/server/game/game.go:
```diff
@@ -0,0 +1,348 @@
+
+	for {
+		select {
+		case <-timerDone:
```
You can just use time.After and drop timerDone channel.
```suggestion
		case <-time.After(time.Duration(g.Timer) * time.Second):
```

bitterfly replied:
This way there is a slight pause between the closing of the ticking channel and the <- time.After because of the creation time. I'm not sure if it's ok.

bitterfly replied:
Done

bitterfly committed:
[backend] Move game outside server folder

Since the game logic should be separate from the server.

bitterfly committed:
[backend] Don't pass argument to lambda

Use it directly instead of passing it as an argument.

bitterfly committed:
[backend] Remove game.NotifyGameInfo

Since the functions doesn't change the game state and only sends event
write directly to the wss whitout using the function.

bitterfly committed:
[backend] Remove database user from game

Create a separate struct for user that has to be sent to the cleint.
The User struct is sent to the client.
The LoginUser struct is used when verifying login/register.

Rename Players -> Words since it now holds only words information.
Separate the players' ids and data into separate struct.

bitterfly committed:
[backend] Make check by Players.IDs instead of Words.ByUser

bitterfly committed:
[backend] Remove unnecessary argument.

bitterfly committed:
[frontend] Don't change state ot tick

Tick event no longer removes the storyteller. Instead it only makes the
timer Nothing and the Storyteller event handles the new storyteller
case.

bitterfly committed:
[backend] Simplify the timer logic

Tick funcions only ticks now (and doesn't send storyteller events).
The MakeTurn functions implements the timeout logic via timer.After
channel and then sends the storyteller event and closes the ticking
channel.

bitterfly committed:
[backend] Remove message package

Move and rename MessageType to EventType in game. Reuse EventType in
server. Prefix all the EventTypes with "Event" for readability.

bitterfly committed:
[backend] Remove the Error channel from game

Since the events are handled on the server if a ws write failes the
server will just log the error. Remove the only other place that writes
in the error channel - send it as an error event instead.

stoyaneft commented:
LGTM
